ddtrace/internal/telemetry/metrics_namespaces.py

```diff
@@ -42,45 +40,36 @@ class MetricNamespace:
     }
 
     def __init__(self):
+        # type: () -> None
         self._flush()
 
     def _flush(self):
+        # type: () -> None
         self._metrics_data = {
-            TELEMETRY_TYPE_GENERATE_METRICS: {TELEMETRY_NAMESPACE_TAG_TRACER: {}, TELEMETRY_NAMESPACE_TAG_APPSEC: {}},
-            TELEMETRY_TYPE_DISTRIBUTION: {TELEMETRY_NAMESPACE_TAG_TRACER: {}, TELEMETRY_NAMESPACE_TAG_APPSEC: {}},
-        }
+            TELEMETRY_TYPE_GENERATE_METRICS: defaultdict(dict),
+            TELEMETRY_TYPE_DISTRIBUTION: defaultdict(dict),
+        }  # type: Dict[str, Dict[str, Dict]]
 
     def get(self):
         # type: () -> Dict[str, Any]
         namespace_metrics = self._metrics_data.copy()
         return namespace_metrics
 
-    @staticmethod
-    def validate_type_metric(metric_check, metric_to_validate):
-        # type: (Metric, Metric) -> bool
-        if metric_check.__class__ == metric_to_validate.__class__:
-            return True
-        return False
-
     def _add_metric(self, metric_type, namespace, name, value=1.0, tags={}, interval=None):
         # type: (MetricType, str,str, float, MetricTagType, Optional[float]) -> None
         """
         Telemetry Metrics are stored in DD dashboards, check the metrics in datadoghq.com/metric/explorer.
         The metric will store in dashboard as "dd.instrumentation_telemetry_data." + namespace + "." + name
         """
-        metric_check = self.metric_class[metric_type](  # type: ignore[abstract]
-            namespace, name, tags=tags, common=True, interval=interval
-        )
+        metric_id = Metric.get_id(name, namespace, tags, metric_type)
         metrics_type_payload = self.metrics_type_payloads[metric_type]
-        # get a Metric that already exists in _metrics data or create a new one (metric_check)
-        metric = self._metrics_data[metrics_type_payload][namespace].get(metric_check.id, metric_check)
-        if not self.validate_type_metric(metric_check, metric):
-            raise TelemetryTypeError(
-                (
-                    'Error: metric with name "%s" and type "%s" exists. You can\'t create a new metric '
-                    'with this name an type "%s"'
-                )
-                % (name, metric.metric_type, metric_check.metric_type)
+
+        existing_metric = self._metrics_data[metrics_type_payload][namespace].get(metric_id)
```

Comment:
We have the following scenarios where we can undercount metrics:
 1. We under count a metric because we update a metric object that has been copied and sent
 2. A metric is queued from different threads at the same time, in this scenario the last value is stored and the first is overwritten.

Since the critical section executes in ~1.4us and we flush metrics every 10 seconds I don't think this is the biggest issue but I will look in alternative approaches. The metrics will not be 100% correct but it should be approximately accurate. I think this is good enough but can probably do better.

Reply:
Looking over how the metrics classes are created and track data, I do think we can get a more thread safe implementation... maybe not entirely lock free, but at least we might be able to reduce the places/scope of what needs a lock.

Reply:
> We under count a metric because we update a metric object that has been copied and sent

we are copying the main dictionary, but all the values are going to be references, so we actually aren't copying the metric classes themselves. we'd need a deep copy for that.


```python
metrics = {}
metrics["a"] = {"b": []}

c = metrics.copy()
c["a"]["b"].append(1)

print(metrics)

# {'a': {'b': [1]}}
```

Reply:
I removed this copy